### ISSUE TITLE
NoMethodError: undefined method `workspace_src_dir' for Chef::Resource::Execute

### ISSUE THREAD

**Issue by rhass** [rhass/ros#6]
This issue is only on the issue-4/catkin-support development branch.
https://github.com/rhass/ros/blob/issue-4/catkin-support/libraries/catkin.rb#L45-L47

I suspect this is an issue in which the resources are not compiled before the method is called, which is resulting in helper provider method returning nil. I tried adding a chef breakpoint to step through it via chef-shell, but it seems breakpoints are not realized until later in the chef lifecycle. I think my next course of action will be to try adding a pry breakpoint.

Nevertheless, I am not sure if I am just not noticing something I am doing completely wrong (most likely), or if this a poise issue (very unlikely) as I have done something very similar to this before in the firefox_package cookbook. https://github.com/rapid7-cookbooks/firefox_package/blob/master/libraries/firefox_package.rb#L66-L77

@coderanger If you can ask for a few minutes to get a nudge in the right direction, it would be greatly appreciated. I am just not sure where I have gone wrong.

```
       Compiling Cookbooks...
       Converging 2 resources
       Recipe: ros_test::install_indigo

       Recipe: apt::default
           * execute[apt-get-update] action run (skipped due to not_if)
           * execute[apt-get update] action nothing (skipped due to action :nothing)
           * execute[apt-get autoremove] action nothing (skipped due to action :nothing)
           * execute[apt-get autoclean] action nothing (skipped due to action :nothing)
        (up to date)
           * execute[apt-get-update-periodic] action run (skipped due to only_if)
           * directory[/var/cache/local] action create (up to date)
           * directory[/var/cache/local/preseeding] action create (up to date)
         Recipe: locale::default
           * apt_package[locales] action install (up to date)
           * execute[Generate locale] action run (skipped due to not_if)
           * execute[Update locale] action run (skipped due to not_if)
           * apt_repository[ros] action add
       * remote_file[/tmp/kitchen/cache/ros.key] action create
        - create new file /tmp/kitchen/cache/ros.key
        - update content in file /tmp/kitchen/cache/ros.key from none to c20cdc
        (new content is binary, diff output suppressed)
        - change mode from '' to '0644'
        (skipped due to not_if)
             * file[/var/lib/apt/periodic/update-success-stamp] action nothing (skipped due to action :nothing)
             * execute[apt-cache gencaches] action nothing (skipped due to action :nothing)
             * execute[apt-get update] action nothing (skipped due to action :nothing)
             * file[/etc/apt/sources.list.d/ros.list] action create (up to date)

           * apt_package[ros-indigo-ros-base] action install (up to date)
           * link[/etc/profile.d/indigo.sh] action create (up to date)
           * execute[rosdep-init] action run (skipped due to not_if)
           - installing ros-indigo-ros-base
       Recipe: ros_test::install_indigo
         * catkin[/opt/catkin_ws] action create
         Recipe: build-essential::_debian
        (up to date)
           * apt_package[binutils-doc] action install (up to date)
           * apt_package[bison] action install (up to date)
           * apt_package[build-essential] action install (up to date)
        (up to date)
           * apt_package[gettext] action install (up to date)
           * apt_package[ncurses-dev] action install (up to date)

           ================================================================================
           Error executing action `create` on resource 'catkin[/opt/catkin_ws]'
           ================================================================================

           NoMethodError
           -------------


           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/ros/libraries/catkin.rb:63:in `block in initialize_workspace'
           /tmp/kitchen/cache/cookbooks/ros/libraries/catkin.rb:62:in `initialize_workspace'
           /tmp/kitchen/cache/cookbooks/ros/libraries/catkin.rb:36:in `action_create'

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/ros_test/recipes/install_indigo.rb

         5: catkin '/opt/catkin_ws' do
             6:   action :create
             7: end

           Compiled Resource:

           # Declared in /tmp/kitchen/cache/cookbooks/ros_test/recipes/install_indigo.rb:5:in `from_file'

           catkin("/opt/catkin_ws") do
             action [:create]
             retries 0
         retry_delay 2
             default_guard_interpreter :default
             declared_type :catkin
             cookbook_name "ros_test"
             recipe_name "install_indigo"
             user "root"
           end


       Running handlers:
       [2015-02-28T08:22:26+00:00] ERROR: Running exception handlers
       Running handlers complete
       [2015-02-28T08:22:26+00:00] ERROR: Exception handlers complete
       [2015-02-28T08:22:26+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       Chef Client failed. 3 resources updated in 38.867048541 seconds
       [2015-02-28T08:22:26+00:00] ERROR: catkin[/opt/catkin_ws] (ros_test::install_indigo line 5) had an error: NoMethodError: undefined method `workspace_src_dir' for Chef::Resource::Execute
       [2015-02-28T08:22:26+00:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
>>>>>> Converge failed on instance <ros-indigo-ubuntu-1404>.
>>>>>> Please see .kitchen/logs/ros-indigo-ubuntu-1404.log for more details
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: SSH exited (1) for command: [sh -c '
sudo -E /opt/chef/bin/chef-client --local-mode --config /tmp/kitchen/client.rb --log_level auto --force-formatter --no-color --chef-zero-port 8889 --json-attributes /tmp/kitchen/dna.json
']
>>>>>> ----------------------
```

**Comment by coderanger** [comment]
Okay, so this is both simple and subtle. The [block on the execute](https://github.com/rhass/ros/blob/issue-4/catkin-support/libraries/catkin.rb#L63-L64) is run in the context of newly created resource object (an instance of `Chef::Execute::Resource`). `Chef::Resource` has a `method_missing` that tries to delegate unknown methods up to the enclosing provider instance, but that doesn't work in this case because the method being called is private and thus not visible.

There are a few options to fix this. The simplest option is to make `workspace_src_dir` public. Another option is a closure inside `initialize_workspace`:

``` ruby
 def initialize_workspace
  workspace = workspace_src_dir
  execute 'catkin_init' do
    command "catkin_init_workspace #{workspace}"
    user new_resource.user
  end
end
```

but the best option is to make `workspace_src_dir` be a public method on the resource class instead. This is generally appropriate for helpers that depend only on the values in the resource and not on system state. You would then call it in the provider via `new_resource.workspace_src_dir` where `#new_resource` follows the delegation mentioned above.

**Comment by rhass** [comment]
Fixed in 62252068e8c614617027b6ce61a1e91c08a26795

The issue turned out to be a bad path reference in which I was attempting to use read an bash script from the path this resource creates, opposed to the general ROS installation path.

https://github.com/rhass/ros/blob/issue-4/catkin-support/libraries/catkin.rb#L31
